### PR TITLE
fix #8993 feat(cirrus): Consistent use of pytest in test cases 🧪 

### LINKED ISSUE #8993: Consitent use of pytest
We have noticed that we are using both pytest and unitests in Fast API, lets convert all of them to pytest to be consitent

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-3592)

### PR DESCRIPTION
**Pull request by yashikakhurana**
Because

- In some test cases we are using `unittest` and in some cases `pytest`

This commit

- Convert all the test cases to use consistent `pytest`
- Removed unnecessary [parameterized](https://pypi.org/project/parameterized/) package and replaces with the pytest parameterization
- Fixtures to reuse object creation

closes #8993

### CODE REVIEW

File: cirrus/server/tests/test_experiment_recipes.py [hunk 1]
```diff
@@ -1,87 +1,77 @@
...
-    def setUp(self):
-        self.sdk = SDK(context)
-
-    def get_remote_settings(self):
-        return RemoteSettings(self.sdk)
-
-    def test_get_recipes_is_empty(self):
-        rs = self.get_remote_settings()
+class TestFetchRecipes:
+    def test_get_recipes_is_empty(self, remote_setting):
```
Comment (jaredlockhart) [inline cirrus/server/tests/test_experiment_recipes.py:8]: Nit: remote settings is pluralized `remote_settings`

File: cirrus/server/tests/test_experiment_recipes.py [hunk 2]
```diff
@@ -1,87 +1,77 @@
...
+        rs = remote_setting
         assert rs.get_recipes() == {"data": []}
 
-    def test_update_recipes(self):
-        rs = self.get_remote_settings()
+    def test_update_recipes(self, remote_setting):
+        rs = remote_setting
```
Comment (jaredlockhart) [inline cirrus/server/tests/test_experiment_recipes.py:13]: Why is this alias necessary?

File: cirrus/server/tests/conftest.py [hunk 3]
```diff
@@ -24,6 +28,26 @@ def remote_setting_mock():
...
+
+
+@fixture
+def fml_setup(fml, sdk):
+    yield fml, sdk
+
+
+@fixture
+def fml():
+    return FeatureManifestLanguage(fml_path, channel)
```
Comment (jaredlockhart) [inline cirrus/server/tests/conftest.py:48]: 🎉 🎉 🎉

**Review by jaredlockhart — COMMENTED** [review]
Couple nits but looks great 🎉 🎉 🎉  But let's wait for @jrbenny35 to review 🙏

File: cirrus/server/tests/test_experiment_recipes.py (hunk 2 above)
Comment (yashikakhurana) [reply cirrus/server/tests/test_experiment_recipes.py:13]: not needed, my brain reading and writing a code making sure doing correctly :D

File: cirrus/server/tests/test_experiment_recipes.py [hunk 4]
```diff
@@ -1,87 +1,77 @@
...
-class TestFetchRecipes(unittest.TestCase):
-    def setUp(self):
-        self.sdk = SDK(context)
-
-    def get_remote_settings(self):
-        return RemoteSettings(self.sdk)
-
-    def test_get_recipes_is_empty(self):
-        rs = self.get_remote_settings()
+class TestFetchRecipes:
```
[the remark this answers is not in the document]
Comment (yashikakhurana) [reply cirrus/server/tests/test_experiment_recipes.py:7]: actually not needed now, just in case we want to share context, I am happy to remove it

**Review by jaredlockhart — APPROVED** [review]
Looks great @yashikakhurana 🎉

File: cirrus/server/tests/test_sdk.py [hunk 5]
```diff
@@ -1,121 +1,132 @@
...
+            "NimbusError.JsonError('JSON Error: missing field `channel`",
+        ),
+    ],
+)
+def test_invalid_context(context, expected_error_message):
+    with pytest.raises(NimbusError) as e:
+        SDK(context=context)
+
+    assert str(e.value).startswith(expected_error_message)
+
```
Comment (jrbenny35) [inline cirrus/server/tests/test_sdk.py:46]: The assert should be indented over.

File: cirrus/server/tests/test_sdk.py [hunk 6]
```diff
@@ -1,121 +1,132 @@
...
+
+def test_failure_case_no_request_context(sdk):
+    targeting_context = {"clientId": "test"}
+
+    try:
+        result = sdk.compute_enrollments(targeting_context)
+    except NimbusError:
+        pytest.fail("NimbusError was raised when it should not have been")
+    assert result == {}
```
Comment (jrbenny35) [inline cirrus/server/tests/test_sdk.py:86]: Could these be parametrized?

File: cirrus/server/tests/test_sdk.py [hunk 7]
```diff
@@ -1,121 +1,132 @@
...
+    assert result == {
+        "enrolledFeatureConfigMap": {},
+        "enrollments": [],
+        "events": [],
+    }
+
+
+def test_set_experiments_failure_missing_data_key(sdk):
+    recipes = '{"invalid": []}'
+
```
Comment (jrbenny35) [inline cirrus/server/tests/test_sdk.py:111]: could you remove this space? Or add it to the other tests in this file that look similar?

File: cirrus/server/tests/test_sdk.py [hunk 8]
```diff
@@ -1,121 +1,132 @@
...
+    assert result == {
+        "enrolledFeatureConfigMap": {},
+        "enrollments": [],
+        "events": [],
+    }
+
+
+def test_set_experiments_failure_malform_key(sdk):
+    recipes = '{"invalid}'
+
```
Comment (jrbenny35) [inline cirrus/server/tests/test_sdk.py:124]: could you remove this space? Or add it to the other tests in this file that look similar?

File: cirrus/server/tests/test_sdk.py [hunk 9]
```diff
@@ -1,121 +1,132 @@
...
+    sdk.set_experiments(recipes)
+    result = sdk.compute_enrollments(targeting_context)
+    assert result == {
+        "enrolledFeatureConfigMap": {},
+        "enrollments": [],
+        "events": [],
+    }
+
+
+def test_set_experiments_failure_missing_data_key(sdk):
```
Comment (jrbenny35) [inline cirrus/server/tests/test_sdk.py:109]: This test and the following look like they could be parametrized

File: cirrus/server/tests/test_feature_manifest.py [hunk 10]
```diff
@@ -1,196 +1,181 @@
...
-            "randomizationUnit": "user_id",
-            "count": 100,
-            "namespace": "",
-            "start": 1,
-            "total": 100,
-        }
+def test_invalid_channel():
+    with pytest.raises(FmlError) as e:
+        FeatureManifestLanguage(fml_path, "invalid_channel")
+    assert str(e.value).startswith("FmlError.InvalidChannelError")
```
Comment (jrbenny35) [inline cirrus/server/tests/test_feature_manifest.py:19]: Does this pass, with the assert outside of the context for where the variable ```e``` is declared?

**Review by jrbenny35 — CHANGES_REQUESTED** [review]
PR #9000! Just some thoughts

File: cirrus/server/tests/test_sdk.py (hunk 6 above)
Comment (yashikakhurana) [reply cirrus/server/tests/test_sdk.py:86]: yes these test will change and all will be parameterised in the upcoming PR

File: cirrus/server/tests/test_feature_manifest.py (hunk 10 above)
Comment (yashikakhurana) [reply cirrus/server/tests/test_feature_manifest.py:19]: it does work, but it should be intended, good point

**Review by jrbenny35 — APPROVED** [review]
pytest rocks

**Pull request merged**